pkg/rpc/rpc.go

```diff
@@ -0,0 +1,166 @@
+package rpc
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"go.uber.org/zap"
+	"io"
+	"sync"
+)
+
+type JsonRpcRequest struct {
+	Method string          `json:"method"`
+	Params json.RawMessage `json:"params"`
+	Id     uint64          `json:"id"`
+}
+
+type JsonRpcResponse struct {
+	Result json.RawMessage `json:"result"`
+	Error  string          `json:"error"`
+	Id     uint64          `json:"id"`
+}
+
+type Error struct {
+	cause error
+}
+
+func (err *Error) Error() string {
+	return fmt.Sprintf("RPC error: %s", err.cause.Error())
+}
+
+func (err *Error) Unwrap() error {
+	return err.cause
+}
+
+type RPC struct {
+	writer    io.Closer // use encoder for writing instead
+	encoder   *json.Encoder
+	encoderMu sync.Mutex
+
+	decoder *json.Decoder
+	logger  *zap.SugaredLogger
+
+	pendingRequests map[uint64]chan JsonRpcResponse
+	lastRequestId   uint64
+	requestsMu      sync.Mutex
+
+	errChannel chan struct{} // never transports a value, only closed through setErr() to signal an occurred error
+	rpcErr     *Error        // only initialised via setErr(), if a rpc (non-skippable) error has occurred
+	errMu      sync.Mutex
+}
+
+// NewRPC creates and returns an RPC object and triggers a go routine to assign incoming responses to given requests.
+// The returned object cannot be used further if an error has occurred and must be recreated.
+func NewRPC(writer io.WriteCloser, reader io.Reader, logger *zap.SugaredLogger) *RPC {
+	rpc := &RPC{
+		writer:          writer,
+		encoder:         json.NewEncoder(writer),
+		decoder:         json.NewDecoder(reader),
+		pendingRequests: map[uint64]chan JsonRpcResponse{},
+		logger:          logger,
+		errChannel:      make(chan struct{}),
+	}
+
+	go rpc.processResponses()
+
+	return rpc
+}
+
+// Call send a request with given parameters and a unique id (to assign responses to requests).
+// Returns the response result or error
+func (r *RPC) Call(method string, params json.RawMessage) (json.RawMessage, error) {
```

Comment:
Again, ID assignment is an implementation detail, these are never visible to the caller.

The description of the returned error should be more verbose here: two different kinds of error can be returned: errors of type `*rpc.Error` mean that the communication failed and that future calls on this instance won't work and a new `*RPC` has to be created, whereas other error just mean that the remote function returned some error (that's non-fatal for the RPC object).